app/assets/javascripts/components/bulkUploadImport.jsx

```diff
@@ -333,131 +380,310 @@ class BulkUploadImport extends React.Component {
+                          </div> : null
+                      }
+                    </div>
+                    <div className='col no-padding s4'>
+                      <button type='button' onClick={this.toggleNewProjectInput}
+                              className='new-project-button new-button skyblue-button tooltipped'
+                              data-position="right" data-delay="50" data-tooltip="Add your desired experiment or project name">
+                        <i className='fa fa-plus'/>
+                        <span>
+                          New project
+                        </span>
+                      </button>
+                    </div>
+                    <div className='col no-padding s12 new-project-input hidden'>
+                      <input type='text' onBlur={ (e) => {
+                        if (e.target.value.trim().length) {
+                          this.handleProjectSubmit();
                         }
-                  </select>
-                  <label>Default host genome</label>
+                        $('.new-project-button').click();
+                      }} ref='new_project' onFocus={ this.clearError } className='browser-default' placeholder='Input new project name' />
+                      {
+                        (this.state.errors.new_project) ?
+                          <div className='field-error'>
+                            {this.state.errors.new_project}
+                          </div> : null
+                      }
+                    </div>
+                  </div>
+                </div>
+                <div className='field'>
+                  <div className='row'>
+                    <div className='col field-title no-padding s5 tooltipped'
+                         data-position="top" data-delay="50"
+                         data-tooltip='This would be subtracted by the pipeline'>
+                      Select host genome
+                    </div>
+                    {
+                      (this.userDetails.admin) ?
+                        <div className='col s7 right-align no-padding right admin-genomes'>
+                          {
+                            this.state.hostGenomes.map((g) => {
+                              if(this.adminGenomes.indexOf(g.name) > 0 && this.userDetails.admin){
+                                 return (
+                                   <div key={g.id}
+                                     className={`${this.state.hostName ===  g.name ? 'active' : ''} genome-label`}
+                                     id={g.name} onClick={() => this.handleHostChange(g.id, g.name)}>
+                                       { g.name }
+                                   </div>
+                                 );
+                              }
+                            })
+                          }
+                        </div> : null
+                    }
+                    </div>
```

Comment:
I think it would be nice to extract the block from line 581 to 598 into a function display_admin_genomes so that it is (a) easier to keep an overview of all the different elements being displayed on the page and (b) easier to figure out which lines to modify if we ever want to change the way admin genomes are displayed going back to this code months from now. Similar comment for lines 603-622 or 535-539

Reply:
right, but it gets harder to keep track of the HTML elements structure if the blocks extracted is pretty long. you won't be able to tell the children element without having to scroll all the way up and down.

Reply:
Hmm ok. Well, maybe it's not too big a deal here, but in general I think it would be a good habit to avoid having too much inline logic scattered around the code, especially repeated logic (e.g. lines 437 and 586). See for example how much good that practice did for the readability of the report page, even with the small disadvantage of not having all html tags in one place